controller: Refactor controller_pod

gaocegege opened the pull request:
Signed-off-by: Ce Gao <gaoce@caicloud.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/tf-operator/548)
<!-- Reviewable:end -->

gaocegege commented:
/assign @ScorpioCPH @jlewi

ScorpioCPH commented on pkg/controller.v2/controller_pod.go:
```diff
@@ -185,23 +116,78 @@ func (tc *TFJobController) reconcilePods(
 
 func getPodSlices(pods []*v1.Pod, replicas int, logger *log.Entry) [][]*v1.Pod {
-	podSlices := make([][]*v1.Pod, 0)
+	logger.Infof("%v", pods)
```
This log maybe too verbose.

ScorpioCPH commented on pkg/controller.v2/controller_pod.go:
```diff
@@ -212,9 +198,9 @@ func getPodSlices(pods []*v1.Pod, replicas int, logger *log.Entry) [][]*v1.Pod {
 		}
```
Should we break here?

gaocegege replied:
Sorry, I rebased the master and can not see it anymore, could you have another review and point it out?

ScorpioCPH replied:
It has already been addressed.

ScorpioCPH commented on pkg/controller.v2/controller_pod.go:
```diff
@@ -185,23 +116,78 @@ func (tc *TFJobController) reconcilePods(
 func getPodSlices(pods []*v1.Pod, replicas int, logger *log.Entry) [][]*v1.Pod {
-	podSlices := make([][]*v1.Pod, 0)
+	logger.Infof("%v", pods)
+	podSlices := make([][]*v1.Pod, replicas)
```
How about use a map for podSlices, like this:
```go
podSlices := make(map[int64][]*v1.Pod, replicas)
```

ScorpioCPH commented on pkg/controller.v2/controller_pod.go:
```diff
@@ -185,23 +116,78 @@ func (tc *TFJobController) reconcilePods(
+	podSlices := make([][]*v1.Pod, replicas)
 	for _, pod := range pods {
 		if _, ok := pod.Labels[tfReplicaIndexLabel]; !ok {
 			logger.Warning("The pod do not have the index label.")
```
continue here?

gaocegege replied:
Added

ScorpioCPH commented on pkg/controller.v2/controller_pod.go:
```diff
@@ -90,84 +84,21 @@ func (tc *TFJobController) reconcilePods(
-		// TODO(CPH): Need to delete pods.
-		loggerForTFJob(tfjob).Infof("need to delete pod but it is not implemented yet")
+		// We already have one, and check if it is succeede or something else.
+		// pod := podSlice[0]
```
What this line used for?

jlewi commented:
nit; it would be great if you could add a more descriptive PR description. If its not making substantive changes I'm happy to leave it to you @ScorpioCPH

jlewi commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/controller.v2/controller_pod.go, line 189 at r2](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAXmP_IgzGMVdAvxccw-r2-189:-LAsgn9KDUAV1Iw5B3T_:bvl4fj) ([raw file](https://github.com/kubeflow/tf-operator/blob/5348a96d600e0ba47f35bf81ac12bcb9aee6afb3/pkg/controller.v2/controller_pod.go#L189)):*
<details><summary><i>Previously, ScorpioCPH (Penghao Cen) wrote…</i></summary><blockquote>

This log maybe too verbose.
</blockquote></details>

IME we should log sufficient information to debug what the job did so we can troubleshoot. If we use the json log entry with appropriate keys; hopefully we can filter logs pretty easily.

---

*[pkg/controller.v2/controller_pod.go, line 46 at r6](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAsi1Nf956RAFbjsPLC:-LAsi1Nf956RAFbjsPLD:b-1h9p72) ([raw file](https://github.com/kubeflow/tf-operator/blob/d6cab712e9c9c25cbefd5974dc2cbc784cd1fb6a/pkg/controller.v2/controller_pod.go#L46)):*
> ```Go
> 	pods = filterPodsForTFReplicaType(pods, rt)
> 	succeeded, failed := getPodStatus(pods)
> 	runningPods := filterRunningPods(pods)
> ```

Should you compute succeeded and running pods in the loop below because you want to take into account the fact that a replica could have more than 1 pod.

Why not use the for loop below to compute a map

replicaIndex -> status  
and then pass that to update status?

---

*[pkg/controller.v2/controller_pod.go, line 53 at r6](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAsh96T2wOgNJOuJH74:-LAsh96T2wOgNJOuJH75:b-l7zmor) ([raw file](https://github.com/kubeflow/tf-operator/blob/d6cab712e9c9c25cbefd5974dc2cbc784cd1fb6a/pkg/controller.v2/controller_pod.go#L53)):*
> ```Go
> 	for index, podSlice := range podSlices {
> 		if len(podSlice) > 1 {
> 			loggerForTFJob(tfjob).Warning("We have to many pods for the worker %d", index)
> ```


More than 1 pod might not be an error. If there is 1 running pod and the rest are not running that might be fine. Might be worth noting that in the TODO as part of fixing this later.

---

*[pkg/controller.v2/controller_pod.go, line 57 at r6](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAshZcTC4NPFxM6fbG5:-LAshZcTC4NPFxM6fbG6:brusd1t) ([raw file](https://github.com/kubeflow/tf-operator/blob/d6cab712e9c9c25cbefd5974dc2cbc784cd1fb6a/pkg/controller.v2/controller_pod.go#L57)):*
> ```Go
> 		}
> 		if len(podSlice) == 0 {
> 			loggerForTFJob(tfjob).Infof("need to create new pod: %s-%d", rt, index)
> ```

Would be good to log this message with keys corresponding to replica type and index. This way we can easily filter log messages to see what we did/didn't do for a particular replica.

---

*[pkg/controller.v2/controller_pod.go, line 138 at r6](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAsgZxn2lRGA4ynPVp5:-LAsgZxn2lRGA4ynPVp6:bku59x3) ([raw file](https://github.com/kubeflow/tf-operator/blob/d6cab712e9c9c25cbefd5974dc2cbc784cd1fb6a/pkg/controller.v2/controller_pod.go#L138)):*
> ```Go
> }
> 
> func getPodSlices(pods []*v1.Pod, replicas int, logger *log.Entry) [][]*v1.Pod {
> ```

Please add a comment for getPodSlices; even thought its a private function it well help future readers of the code.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubeflow/tf-operator/548)*
<!-- Sent from Reviewable.io -->

gaocegege commented:
*[pkg/controller.v2/controller_pod.go, line 138 at r6](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAsgZxn2lRGA4ynPVp5:-LAuVS_k-GMAgM7li-FJ:bpwpp9e) ([raw file](https://github.com/kubeflow/tf-operator/blob/d6cab712e9c9c25cbefd5974dc2cbc784cd1fb6a/pkg/controller.v2/controller_pod.go#L138)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Please add a comment for getPodSlices; even thought its a private function it well help future readers of the code.
</blockquote></details>

OK, SGTM

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubeflow/tf-operator/548)*
<!-- Sent from Reviewable.io -->

gaocegege commented:
*[pkg/controller.v2/controller_pod.go, line 189 at r2](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAXmP_IgzGMVdAvxccw-r2-189:-LAuVZzT2lUyIYIidXMb:bxc7diw) ([raw file](https://github.com/kubeflow/tf-operator/blob/5348a96d600e0ba47f35bf81ac12bcb9aee6afb3/pkg/controller.v2/controller_pod.go#L189)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

IME we should log sufficient information to debug what the job did so we can troubleshoot. If we use the json log entry with appropriate keys; hopefully we can filter logs pretty easily.
</blockquote></details>

I added such statement for debug and removed now. I think the log in this function is enough to help us locate the errors. WDYT

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubeflow/tf-operator/548)*
<!-- Sent from Reviewable.io -->

gaocegege commented:
*[pkg/controller.v2/controller_pod.go, line 190 at r2](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAXmP_IgzGMVdAvxccw-r2-190:-LAuW69s2T7CoWc-YhGD:b-p0o176) ([raw file](https://github.com/kubeflow/tf-operator/blob/5348a96d600e0ba47f35bf81ac12bcb9aee6afb3/pkg/controller.v2/controller_pod.go#L190)):*
<details><summary><i>Previously, ScorpioCPH (Penghao Cen) wrote…</i></summary><blockquote>

How about use a map for podSlices, like this:
```go
podSlices := make(map[int64][]*v1.Pod, replicas)
``` 
</blockquote></details>

I considered it once but I think the index is in a range then I think slice is easier to understand.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubeflow/tf-operator/548)*
<!-- Sent from Reviewable.io -->

gaocegege commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.io/reviews/kubeflow/tf-operator/548#-:-LAuWdO715Lvskw4SYqi:bnfp4nl)*
<!-- Sent from Reviewable.io -->

gaocegege commented:
*[pkg/controller.v2/controller_pod.go, line 53 at r6](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAsh96T2wOgNJOuJH74:-LAuWuoBEsmAfDGq3ojr:brvsc7e) ([raw file](https://github.com/kubeflow/tf-operator/blob/d6cab712e9c9c25cbefd5974dc2cbc784cd1fb6a/pkg/controller.v2/controller_pod.go#L53)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

More than 1 pod might not be an error. If there is 1 running pod and the rest are not running that might be fine. Might be worth noting that in the TODO as part of fixing this later.
</blockquote></details>

I think it is an error since we expect that only one pod for one replica index is created.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubeflow/tf-operator/548)*
<!-- Sent from Reviewable.io -->

gaocegege commented:
*[pkg/controller.v2/controller_pod.go, line 46 at r6](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAsi1Nf956RAFbjsPLC:-LAuX4bd0STUnkwq76uK:b-ssu2x2) ([raw file](https://github.com/kubeflow/tf-operator/blob/d6cab712e9c9c25cbefd5974dc2cbc784cd1fb6a/pkg/controller.v2/controller_pod.go#L46)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Should you compute succeeded and running pods in the loop below because you want to take into account the fact that a replica could have more than 1 pod.

Why not use the for loop below to compute a map

replicaIndex -> status  
and then pass that to update status?

</blockquote></details>

Good idea, I will update it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubeflow/tf-operator/548)*
<!-- Sent from Reviewable.io -->

gaocegege commented:
*[pkg/controller.v2/controller_pod.go, line 57 at r6](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAshZcTC4NPFxM6fbG5:-LAuXE4w78Pvr2ust9xJ:bman693) ([raw file](https://github.com/kubeflow/tf-operator/blob/d6cab712e9c9c25cbefd5974dc2cbc784cd1fb6a/pkg/controller.v2/controller_pod.go#L57)):*
<details><summary><i>Previously, jlewi (Jeremy Lewi) wrote…</i></summary><blockquote>

Would be good to log this message with keys corresponding to replica type and index. This way we can easily filter log messages to see what we did/didn't do for a particular replica.
</blockquote></details>

OK, will do it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubeflow/tf-operator/548)*
<!-- Sent from Reviewable.io -->

gaocegege commented:
*[pkg/controller.v2/controller_pod.go, line 101 at r2](https://reviewable.io/reviews/kubeflow/tf-operator/548#-LAXmP_IgzGMVdAvxccw-r2-101:-LAuXJSLA9NFm9omOGO6:bss4a2b) ([raw file](https://github.com/kubeflow/tf-operator/blob/5348a96d600e0ba47f35bf81ac12bcb9aee6afb3/pkg/controller.v2/controller_pod.go#L101)):*
<details><summary><i>Previously, ScorpioCPH (Penghao Cen) wrote…</i></summary><blockquote>

What this line used for?
</blockquote></details>

Removed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubeflow/tf-operator/548)*
<!-- Sent from Reviewable.io -->

ScorpioCPH commented on pkg/controller.v2/controller_status.go:
```diff
@@ -0,0 +1,66 @@
+package controller
```
Please add copyright here :)

gaocegege replied:
Sorry for that. I will add it soon.

ScorpioCPH commented on pkg/controller.v2/controller_status.go:
```diff
@@ -0,0 +1,66 @@
+		}
+	}
+
+	// All workers are running, set StartTime
```
nit: add `.`

gaocegege replied:
OK

gaocegege commented:
PTAL @ScorpioCPH

yph152 commented:
```
 else {
 +			// We already have one, and check the status.
 +			pod := podSlice[0]
 +			updateTFJobReplicaStatuses(tfjob, rtype, pod)
  		}		  		}
...
func updateTFJobReplicaStatuses(tfjob *tfv1alpha2.TFJob, rtype tfv1alpha2.TFReplicaType, pod *v1.Pod) {
 +	switch pod.Status.Phase {
 +	case v1.PodRunning, v1.PodPending:
 +		tfjob.Status.TFReplicaStatuses[rtype].Active++
 +	case v1.PodSucceeded:
 +		tfjob.Status.TFReplicaStatuses[rtype].Succeeded++
 +	case v1.PodFailed:
 +		tfjob.Status.TFReplicaStatuses[rtype].Failed++
 +	}
 +}
```
@gaocegege If the state of pod continues to change, it will always be restarted, and the state of
```running --> failed --> running --> failed...```, These values are going to get bigger all the time.

gaocegege commented:
@yph152 I do not think so, because of https://github.com/kubeflow/tf-operator/pull/548/files#diff-a5453ed67196062408fa782a70f4be38R47

yph152 commented:
thanks，I understand！

ScorpioCPH commented:
/lgtm
/approve
